pkg/provider/hub/handler.go

```diff
@@ -0,0 +1,147 @@
+package hub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync/atomic"
+
+	"github.com/traefik/traefik/v2/pkg/config/dynamic"
+	"github.com/traefik/traefik/v2/pkg/log"
+)
+
+type handler struct {
+	mux *http.ServeMux
+
+	client http.Client
+
+	entryPoint string
+	port       int
+	tlsCfg     *TLS
+
+	// Accessed atomically.
+	lastCfgUnixNano int64
+
+	cfgChan chan<- dynamic.Message
+}
+
+func newHandler(entryPoint string, port int, cfgChan chan<- dynamic.Message, tlsCfg *TLS, client http.Client) http.Handler {
+	h := &handler{
+		mux:        http.NewServeMux(),
+		entryPoint: entryPoint,
+		port:       port,
+		cfgChan:    cfgChan,
+		tlsCfg:     tlsCfg,
+		client:     client,
+	}
+
+	h.mux.HandleFunc("/config", h.handleConfig)
+	h.mux.HandleFunc("/discover-ip", h.handleDiscoverIP)
+	h.mux.HandleFunc("/state", h.handleState)
+
+	return h
+}
+
+type configRequest struct {
+	UnixNano      int64                  `json:"unixNano"`
+	Configuration *dynamic.Configuration `json:"configuration"`
+}
+
+func (h *handler) handleConfig(rw http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		http.Error(rw, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return
+	}
+
+	payload := &configRequest{Configuration: emptyDynamicConfiguration()}
+	if err := json.NewDecoder(req.Body).Decode(payload); err != nil {
+		err = fmt.Errorf("decodig config request: %w", err)
+		log.WithoutContext().Errorf("Handling config: %v", err)
+		http.Error(rw, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+
+	atomic.StoreInt64(&h.lastCfgUnixNano, payload.UnixNano)
```

Comment:
We record that this is the last update we did, but actually do not guarantee that it is true, since we may drop the message below if traefik is not ready to receive it.

so I think we when serve the /state , we potentially are going to lie, because we might claim that the an update was done at time t, but the update at time t was actually dropped because traefik was not ready for it. no?